### PR TITLE
shell: Add support for asynchronous UART API via uart_async_rx helper

### PR DESCRIPTION
**Pull request by nordic-krch**
Rework UART backend to clearly use single instance approach as before it was a bit messy with some parts of implementation indicating multi-instance approach and some single instance. Backend has been around for years and multi-instance requirement never came. 

Added support for UART asynchronous API which is more efficient in terms of power consumption and performance. 

Support for asynchronous UART API is done using `uart_async_rx` helper module which is also added in this PR. It is extracted from shell backend because there is a second use case on horizon that will soon come (uart interrupt driven to async adaptation layer). `uart_async_rx` helps to handle reception using asynchronous API in cases where incoming data has variable length and is random.

### CODE REVIEW

File: drivers/serial/Kconfig [hunk 1]
```diff
@@ -125,6 +125,14 @@ config UART_PIPE
 	  data (as contrary to console UART driver) and all aspects of received
 	  protocol data are handled by application provided callback.
 
+config UART_ASYNC_RX
```
Comment (dcpleung) [inline drivers/serial/Kconfig:128]: `UART_ASYNC_RX_HELPER` or `UART_ASYNC_RX_AGG` (aggregator) or `UART_ASYNC_RX_POOL` (receiving pool but will definitely cause confusion with polling)?

Comment (nordic-krch) [reply drivers/serial/Kconfig]: Changed to `UART_ASYNC_RX_HELPER`. I did not go with renaming files from `uart_async_rx.[ch]` to `uart_async_rx_helper.[ch]`. What do you think? If renaming the file what about API calls? adding `_helper_` part will make names quite long.

Comment (dcpleung) [reply drivers/serial/Kconfig]: Hm... how about naming the whole thing as "buffered RX"? So the prefix could be `uart_async_buf_rx_`. I think this is more clear on its purpose than simple "helper".

File: drivers/serial/Kconfig [hunk 2]
```diff
@@ -125,6 +125,14 @@ config UART_PIPE
 	  data (as contrary to console UART driver) and all aspects of received
 	  protocol data are handled by application provided callback.
 
+config UART_ASYNC_RX_HELPER
+	bool "Helper for UART asynchronous reception"
+	help
+	  Module implements handling of reception of variable length data using
```
Comment (jakub-uC) [inline drivers/serial/Kconfig:131]: I suggest to tune it a bit:

```
The module implements the handling of variable-length data reception using the Asynchronous UART API.
It is suitable for scenarios where there may be a delay in processing received data. 
The module utilizes a zero-copy approach with multiple reception buffers.
```

File: drivers/serial/uart_async_rx.c [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+	uint8_t *p = rx_data->config->buffer;
+
+	p += idx * (rx_data->buf_len + sizeof(struct uart_async_rx_buf));
+
+	return (struct uart_async_rx_buf *)p;
+}
+
+uint8_t *uart_async_rx_buf_req(struct uart_async_rx *rx_data)
+{
+	uint8_t *data = NULL;
```
Comment (jakub-uC) [inline drivers/serial/uart_async_rx.c:27]: `__ASSERT_NO_MSG(rx_data)`

File: drivers/serial/uart_async_rx.c [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+		atomic_dec(&rx_data->free_buf_cnt);
+	}
+
+	return data;
+}
+
+void uart_async_rx_on_rdy(struct uart_async_rx *rx_data, uint8_t *buffer, size_t length)
+{
+	/* Cannot use CONTAINER_OF because validation fails due to type mismatch:
```
Comment (jakub-uC) [inline drivers/serial/uart_async_rx.c:43]: I have a small suggestion. For API functions, it's worth adding `__ASSERT_NO_MSG` to check pointers.

**Review by jakub-uC — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Comment by ycsin** [comment]
@nordic-krch this seems to have caused the CI to fail: https://github.com/zephyrproject-rtos/zephyr/actions/runs/6870450167/job/18685447368

```
/__w/zephyr/zephyr/subsys/mgmt/mcumgr/transport/src/smp_shell.c: In function 'smp_shell_input_timeout_handler':
/__w/zephyr/zephyr/subsys/mgmt/mcumgr/transport/src/smp_shell.c:68:53: error: invalid use of undefined type 'struct shell_uart'
   68 |         struct smp_shell_data *const data = &sh_uart->ctrl_blk->smp;
      |     
```

**Comment by ycsin** [comment]
This patch is breaking our internal firmware which depended on the previous `shell_uart`'s multi-instance components, cc @cfriedt

**Comment by nordic-krch** [comment]
@ycsin can you provide more details? Previously shell_uart was single instance with multi-instance notion. I clarified it to be single-instance since in 5 years I never heard of anyone asking for multi-instance. It seems that I've been wrong.

**Comment by ycsin** [comment]
> @ycsin can you provide more details?

previously it was possible to create another shell UART instance (in another file) with `SHELL_UART_DEFINE` + `SHELL_DEFINE`, to reuse almost everything (`shell_uart_transport_api`, ...) in `shell_uart.c`, since the buffer is created by `SHELL_UART_DEFINE`

**Pull request merged**